nvm.sh

```diff
@@ -275,6 +275,357 @@ nvm_find_up() {
+          else
+            node_version_compatible=$FALSE
+            no_remote_version_will_satisfy_all_comparators=$TRUE
+          fi
+
+        elif nvm_string_contains_regexp "$comparator" '^<[0-9]+\.[0-9]+\.[0-9]+$'; then
+          if nvm_version_greater "$stripped_version_from_comparator" "$current_remote_node_version"; then
+            node_version_compatible=$TRUE
+          else
+            node_version_compatible=$FALSE
+          fi
+
+        elif nvm_string_contains_regexp "$comparator" '^>[0-9]+\.[0-9]+\.[0-9]+$'; then
+          if nvm_version_greater "$current_remote_node_version" "$stripped_version_from_comparator"; then
+            node_version_compatible=$TRUE
+          else
+            node_version_compatible=$FALSE
+            no_remote_version_will_satisfy_all_comparators=$TRUE
+          fi
+
+        else
+          node_version_compatible=$FALSE
+        fi
+        # stop checking if all comparators are compatible with current_remote_node_version upon finding one that is incompatible
+        [ $node_version_compatible -eq $FALSE ] && comparator_set_copy=''
+      done # while [ -n "$comparator_set_copy" ]; do
+      # stop iterating through remote_node_versions_copy for this comparator_set upon finding the first remote node version that is compatible with all comparators.
+      [ $node_version_compatible -eq $TRUE ] && highest_compatible_node_versions="$highest_compatible_node_versions $current_remote_node_version"
+    done # while [ -n "$remote_node_versions_copy" ] && [ $node_version_compatible -eq $FALSE ] && [ $no_remote_version_will_satisfy_all_comparators -eq $FALSE ]; do
+
+  done
+
+  highest_compatible_node_version='0.0.0'
+
+  highest_compatible_node_versions=$(command printf "%s" "$highest_compatible_node_versions" | command tr ' ' '\n')
+  while [ -n "$highest_compatible_node_versions" ]; do
+    compatible_node_version=$(command printf "%s" "$highest_compatible_node_versions" | head -n1 | command sed -E 's/^ +//;s/ +$//')
+    highest_compatible_node_versions=$(command printf "%s" "$highest_compatible_node_versions" | tail -n +2)
+    [ -n "$compatible_node_version" ] || continue
+
+    if nvm_version_greater "$compatible_node_version" "$highest_compatible_node_version"; then
+      highest_compatible_node_version="$compatible_node_version"
+    fi
+  done
+  [ "$highest_compatible_node_version" != '0.0.0' ] && nvm_echo "$highest_compatible_node_version"
+) }
+
+nvm_find_package_json() { (
+  dir="$(nvm_find_up 'package.json')"
+  if [ -e "${dir}/package.json" ]; then
+    nvm_echo "${dir}/package.json"
+  fi
+) }
+
+# extracts engines.node value from package.json exactly except:
+#  - removes all line breaks and carriage returns
+#  - normalizes all consecutive whitespace to 1 occurrence
+#  - must match regexp: "[|<> [:alnum:].^=~*-]\+"
+nvm_get_node_from_pkg_json() { (
+  package_json_contents=${1-}
+  engines_node_value=''
+  open_brackets=0
+  closed_brackets=0
+  in_quotes='false'
```

Comment:
any reason not to use 0/1 for these?

Reply:
It would be a unclear that `in_quotes` stores a binary value, and `open_brackets` and `closed_brackets` store a counter value. But maybe that problem could be mitigated by just adding comments.